### PR TITLE
test(CNV-43603): add disk-uploader functional tests

### PR DESCRIPTION
**Pull request by codingben**
**What this PR does / why we need it**:

Deploy a new Cirros VM to export from,
and run disk-uploader as Task to execute
extraction process and upload it to the
container registry.

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by codingben** [comment]
@ksimon1 @0xFelix Can you please take a look?

File: test/disk_uploader_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+		dvName := constants.E2ETestsRandomName("example-dv")
+		secretName := constants.E2ETestsRandomName("disk-uploader-credentials")
+
+		config := &testconfigs.DiskUploaderTestConfig{
+			TaskRunTestConfig: testconfigs.TaskRunTestConfig{},
+			TaskData: testconfigs.DiskUploaderTaskData{
+				ExportSourceKind: "vm",
+				ExportSourceName: vmName,
+				VolumeName:       dvName,
+				ImageDestination: "quay.io/boukhano/e2e-example-vm", // TODO: Use quay.io/kubevirt/e2e-example-vm
```
Comment (0xFelix) [inline test/disk_uploader_test.go:47]: Is there a registry local to the test cluster, that can be used?

File: test/disk_uploader_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+			"accessKeyId": []byte(""), // TODO: Get value from OCP CI
+			"secretKey":   []byte(""), // TODO: Get value from OCP CI
+		},
+	}
+
+	_, err := client.Secrets(namespace).Create(context.Background(), v1Secret, metav1.CreateOptions{})
+	Expect(err).ShouldNot(HaveOccurred())
+}
+
+func CreateAndWaitForVM(client kubevirtcliv1.KubevirtClient, namespace, name, dataVolumeName string, timeout time.Duration) {
```
Comment (0xFelix) [inline test/disk_uploader_test.go:80]:
```suggestion
func createAndWaitForVM(client kubevirtcliv1.KubevirtClient, namespace, name, dataVolumeName string, timeout time.Duration) {
```

File: test/disk_uploader_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	v1VirtualMachine := &kubevirtv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: kubevirtv1.VirtualMachineSpec{
```
Comment (0xFelix) [inline test/disk_uploader_test.go:85]: Since `kubevirt.io/kubevirt` is already in the dependencies of this project, maybe you can use `libvmi` to shorten this?

File: test/disk_uploader_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+func createRegistryCredentials(client kubeclientcorev1.CoreV1Interface, namespace, name string) {
+	v1Secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			"accessKeyId": []byte(""), // TODO: Get value from OCP CI
+			"secretKey":   []byte(""), // TODO: Get value from OCP CI
```
Comment (0xFelix) [inline test/disk_uploader_test.go:72]: Please don't add any credentials to this repo! Can we do this differently with a local registry?

File: test/disk_uploader_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		f.TestSetup(config)
+
+		createRegistryCredentials(f.CoreV1Client, f.DeployNamespace, secretName)
+		CreateAndWaitForVM(f.KubevirtClient, f.DeployNamespace, vmName, dvName, config.GetTaskRunTimeout())
+
+		runner.NewTaskRunRunner(f, config.GetTaskRun()).
+			CreateTaskRun().
+			ExpectSuccess().
+			ExpectLogs(config.GetAllExpectedLogs()...).
+			ExpectResults(nil)
```
Comment (0xFelix) [inline test/disk_uploader_test.go:60]: Query the repository after to verify disk was uploaded?

File: test/disk_uploader_test.go (hunk 4 above)
Comment (ksimon1) [reply test/disk_uploader_test.go:72]: There will be no credentials in the repo, since the OCP ci has a secret vault, which can be used for populating the secrets in test

File: test/disk_uploader_test.go (hunk 1 above)
Comment (codingben) [reply test/disk_uploader_test.go:47]: I had a discussion with @ksimon1 about it to keep using an external registry that allows you also to test it on Kubernetes cluster. As you know, only OpenShift has built-in container registry.

File: test/disk_uploader_test.go (hunk 3 above)
Comment (codingben) [reply test/disk_uploader_test.go:85]: Yes I was thinking about this. I'll check the usage of it.

File: test/disk_uploader_test.go (hunk 4 above)
Comment (codingben) [reply test/disk_uploader_test.go:72]: There will be no secrets in this repository. Both `accessKeyId` and `secretKey` values will be set from the environment variables (aka Secrets in the OCP CI).

File: test/disk_uploader_test.go (hunk 5 above)
Comment (codingben) [reply test/disk_uploader_test.go:60]: Yep good idea, need to check it.

File: test/disk_uploader_test.go (hunk 2 above)
Comment (codingben) [reply test/disk_uploader_test.go:80]: I need to start wear glasses :smile:

**Comment by codingben** [comment]
@0xFelix @ksimon1 This [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/eb6a2ad705868f2a24169b1953bc498b9f8675e0..e24fe1af435d9bdfcabd8768ca80f92391f65475) adds the usage of libvmi and libdv. CI or developer locally expected to have `REGISTRY_IMAGE_DESTINATION`, `REGISTRY_ACCESS_KEY_ID` and `REGISTRY_SECRET_KEY` environment variables.

**Comment by codingben** [comment]
@ksimon1 @0xFelix Currently there is a single test case: `Extracts disk from VM and upload to container registry`. Do you think we should also test scenarios of extracting it from VMSnapshot or PVC, or it's enough to have only VM? I don't want to end up re-testing and over-testing Export API.

File: test/disk_uploader_test.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	_, err := client.Secrets(namespace).Create(context.Background(), v1Secret, metav1.CreateOptions{})
+	Expect(err).ToNot(HaveOccurred())
+}
+
+func createAndWaitForVM(client kubevirtcliv1.KubevirtClient, namespace, name string, timeout time.Duration) {
+	dataVolume := libdv.NewDataVolume(
+		libdv.WithRegistryURLSource("docker://quay.io/kubevirt/cirros-container-disk-demo"),
+		libdv.WithForceBindAnnotation(),
```
Comment (0xFelix) [inline test/disk_uploader_test.go:85]: Should not be required, if the VM was started at least once.

File: test/disk_uploader_test.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	)
+	v1VirtualMachine := libvmi.NewVirtualMachine(
+		libvmi.New(
+			libvmi.WithDataVolume("datavolumedisk", dataVolume.Name),
+			libvmi.WithResourceMemory("5Gi"),
```
Comment (0xFelix) [inline test/disk_uploader_test.go:90]: That's a huge amount of memory for a cirros VM.

File: test/disk_uploader_test.go (hunk 6 above)
Comment (codingben) [reply test/disk_uploader_test.go:85]: VM can't be exported if started and it's running. It should be stopped.

File: test/disk_uploader_test.go (hunk 7 above)
Comment (codingben) [reply test/disk_uploader_test.go:90]: Yeah will reduce it.

File: test/disk_uploader_test.go (hunk 6 above)
Comment (codingben) [reply test/disk_uploader_test.go:85]: It's created as stopped by default.

**Comment by codingben** [comment]
@ksimon1 @0xFelix This [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/e24fe1af435d9bdfcabd8768ca80f92391f65475..ef903df540d552fc6056d927f6fd74fb39e95f29) will make e2e tests work, as tested locally it works now and image was pushed to the Quay container registry [1].

Can you please review?

[1] https://quay.io/repository/boukhano/e2e-tests-example-vm

File: automation/e2e-deploy-resources.sh [hunk 8]
```diff
@@ -40,7 +40,7 @@ kubectl apply -f "https://github.com/kubevirt/kubevirt/releases/download/${KUBEV
 
 kubectl apply -f "https://github.com/kubevirt/kubevirt/releases/download/${KUBEVIRT_VERSION}/kubevirt-cr.yaml"
 
-kubectl patch kubevirt kubevirt -n kubevirt --type merge -p '{"spec":{"configuration":{"developerConfiguration":{"featureGates": ["DataVolumes"]}}}}'
+kubectl patch kubevirt kubevirt -n kubevirt --type merge -p '{"spec":{"configuration":{"developerConfiguration":{"featureGates": ["DataVolumes", "VMExport"]}}}}'
```
Comment (0xFelix) [inline automation/e2e-deploy-resources.sh:43]: `DataVolumes` FG can be dropped I guess.

File: automation/e2e-source.sh [hunk 9]
```diff
@@ -3,3 +3,6 @@
 export DEV_MODE="${DEV_MODE:-false}"
 export STORAGE_CLASS="${STORAGE_CLASS:-}"
 export NUM_NODES=${NUM_NODES:-2}
+export REGISTRY_IMAGE_DESTINATION="quay.io/kubevirt/e2e-tests-example-vm"
```
Comment (0xFelix) [inline automation/e2e-source.sh:6]: Is there absolutely no way to use an ephemeral registry? Why do we have to store test artifacts on a live registry? (quay.io)

File: test/disk_uploader_test.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+		Expect(imageDestination).ToNot(BeEmpty())
+
+		registryCredentials, err := createRegistryCredentials(f.CoreV1Client, f.DeployNamespace)
+		Expect(err).ToNot(HaveOccurred())
+
+		cirrosVm, err := createCirrosVM(f.KubevirtClient, f.DeployNamespace)
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = vm.WaitForVM(f.KubevirtClient, f.DeployNamespace, cirrosVm.Name, "", constants.Timeouts.WaitForVMStart.Duration, false)
+		Expect(err).ToNot(HaveOccurred())
```
Comment (0xFelix) [inline test/disk_uploader_test.go:48]: Move this to BeforeEach?

File: test/disk_uploader_test.go (hunk 5 above)
Comment (0xFelix) [reply test/disk_uploader_test.go:66]: Are you going to add a check?

File: test/disk_uploader_test.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+			"secretKey":   []byte(registryKey),
+		},
+	}
+
+	return client.Secrets(namespace).Create(context.Background(), v1Secret, metav1.CreateOptions{})
+}
+
+func createCirrosVM(client kubevirtcliv1.KubevirtClient, namespace string) (*v1.VirtualMachine, error) {
+	dataVolume := libdv.NewDataVolume(
+		libdv.WithRegistryURLSource("docker://quay.io/kubevirt/cirros-container-disk-demo"),
```
Comment (0xFelix) [inline test/disk_uploader_test.go:93]: Please do not use `libdv`. It lives under `tests/libdv` which might be subject to change. Only `libvmi` can be considered stable.

File: automation/e2e-source.sh (hunk 9 above)
Comment (codingben) [reply automation/e2e-source.sh:6]: I replied [here](https://github.com/kubevirt/kubevirt-tekton-tasks/pull/549#discussion_r1839821263) earlier. The issue is that there is no built-in registry in Kubernetes cluster (only on OpenShift).

So we have 3 options:

1). Deploy a new pod that will use [registry image](https://hub.docker.com/_/registry) to push there
2). Use OCP's built-in container registry
3). Use an external container registry

@0xFelix @ksimon1 What do you prefer?

File: test/disk_uploader_test.go (hunk 11 above)
Comment (codingben) [reply test/disk_uploader_test.go:93]: @0xFelix Don't you think that this is an issue that such important libs for external projects tests living under `kubevirt/kubevirt`, especially under `tests/libdv`?

File: automation/e2e-deploy-resources.sh (hunk 8 above)
Comment (codingben) [reply automation/e2e-deploy-resources.sh:43]: Yep just want to get confirmation from @ksimon1 about it, so this PR won't do any other changes except adding new tests.

File: test/disk_uploader_test.go (hunk 5 above)
Comment (codingben) [reply test/disk_uploader_test.go:66]: Yes. We may want to delete the repository after completing running the tests. If we won't delete it, and re-run the functional tests, it may be false check.

**Comment by codingben** [comment]
> @ksimon1 @0xFelix Currently there is a single test case: `Extracts disk from VM and upload to container registry`. Do you think we should also test scenarios of extracting it from VMSnapshot or PVC, or it's enough to have only VM? I don't want to end up re-testing and over-testing [Export API](https://kubevirt.io/user-guide/storage/export_api).

@ksimon1 @0xFelix What is your opinion about it?

File: test/disk_uploader_test.go (hunk 11 above)
Comment (codingben) [reply test/disk_uploader_test.go:93]: @0xFelix @ksimon1 I raised this discussion in the code quality meeting. I got an approval from @akalenyu to relocate `libdv` from `tests` to `pkg` - needs to see how it'll work.

File: automation/e2e-deploy-resources.sh (hunk 8 above)
Comment (ksimon1) [reply automation/e2e-deploy-resources.sh:43]: yes, it can be dropped

File: test/disk_uploader_test.go (hunk 11 above)
Comment (codingben) [reply test/disk_uploader_test.go:93]: Done.

Comment (codingben) [reply test/disk_uploader_test.go]: This [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/ef903df540d552fc6056d927f6fd74fb39e95f29..8350664b97acdf4964327a2477387254d29d611a) removes the usage of `tests/libdv` from `kubevirt/kubevirt` repository.

File: automation/e2e-deploy-resources.sh (hunk 8 above)
Comment (codingben) [reply automation/e2e-deploy-resources.sh]: I dropped it.

File: test/disk_uploader_test.go (hunk 10 above)
Comment (codingben) [reply test/disk_uploader_test.go]: Why? this is a specific initialization of a VM for this test case, we don't have more test cases that can benefit from it at the moment. Also, it would require to introduce variables outside this scope.

File: automation/e2e-source.sh [hunk 12]
```diff
@@ -3,3 +3,6 @@
 export DEV_MODE="${DEV_MODE:-false}"
 export STORAGE_CLASS="${STORAGE_CLASS:-}"
 export NUM_NODES=${NUM_NODES:-2}
+export IN_CLUSTER_IMAGE_REGISTRY="image-registry.openshift-image-registry.svc:5000"
+export IN_CLUSTER_REGISTRY_ACCESS_KEY_ID=""
+export IN_CLUSTER_REGISTRY_SECRET_KEY=""
```
Comment (codingben) [inline automation/e2e-source.sh:8]: @0xFelix @ksimon1 It seems like OCP's in-cluster registry requires authorization (returns 401 if username/password not set).

File: test/disk_uploader_test.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+			Name: constants.E2ETestsRandomName("disk-uploader-credentials"),
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			"accessKeyId": []byte(registryKeyId),
+			"secretKey":   []byte(registryKey),
+		},
+	}
+
+	return client.Secrets(namespace).Create(context.Background(), v1Secret, metav1.CreateOptions{})
```
Comment (codingben) [inline test/disk_uploader_test.go:89]: @0xFelix @ksimon1 This secret is still required even with OCP's in-cluster registry. If there is no secret, and no secret-name passed to the Tekton's task run, then it would fail to get the `accessKeyId` and `secretKey` from `secretKeyRef` [1].

[1] https://github.com/kubevirt/kubevirt-tekton-tasks/blob/main/release/tasks/disk-uploader/disk-uploader.yaml#L48-L57

Comment (ksimon1) [reply test/disk_uploader_test.go:89]: But in this case it should not be a problem to store credentials inside the tests, since the internal registry of the test cluster is accessible only from within the cluster and not from outside.

File: automation/e2e-source.sh (hunk 12 above)
Comment (ksimon1) [reply automation/e2e-source.sh:8]: have you tried https://access.redhat.com/solutions/6959306?

File: test/disk_uploader_test.go (hunk 5 above)
Comment (codingben) [reply test/disk_uploader_test.go:67]: > Are you going to add a check?

I'm not sure if it's needed. Pushing the image using go-containerregistry seems like already verifies if it was pushed successfully or not, and if not, then it's trying to retry it [1].

[1] https://github.com/google/go-containerregistry/blob/main/pkg/v1/remote/write.go#L620

Comment (ksimon1) [reply test/disk_uploader_test.go:67]: But if image already exists, the disk-uploader will repush the image and succeed.
However, these tests runs always in different namespace, and AFAIR images pushed with namespace in its name, will be treatet as a new image.

**Comment by codingben** [comment]
Functional tests locally passed, marked this PR as ready to trigger the CI tests.

**Comment by codingben** [comment]
e2e-tests job failed with this error:
```
+++ oc whoami --show-token
error: no token is currently in use for this session
```
I'll take a look and will push to resolve it.

File: modules/disk-uploader/pkg/parse/clioptions.go
Comment (0xFelix) [inline modules/disk-uploader/pkg/parse/clioptions.go:1]: Put these changes into a separate commit?

File: test/disk_uploader_test.go (hunk 5 above)
Comment (0xFelix) [reply test/disk_uploader_test.go:67]: Still, isn't the point of functional tests to verify that your task did what it says it did?

Comment (ksimon1) [reply test/disk_uploader_test.go:67]: For sure I agree, I was responding to the part about the cleanup :)

Comment (codingben) [reply test/disk_uploader_test.go:67]: @0xFelix I'm working on checking it, will push it soon.

**Comment by codingben** [comment]
@0xFelix @ksimon1 This [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/900b3a103dfc4abb373a712919450c6ee1edd83c..c6906fca2c968be16be011b295adbda947b9f464) will check if the containerDisk exists in the container registry. Can you please review?

**Comment by codingben** [comment]
From e2e-tests job logs:
```
++ IN_CLUSTER_IMAGE_REGISTRY=image-registry.openshift-image-registry.svc:5000/e2e-tests-41458/e2e-exported-disk:latest
+++ oc whoami
++ export IN_CLUSTER_REGISTRY_ACCESS_KEY_ID=system:admin
++ IN_CLUSTER_REGISTRY_ACCESS_KEY_ID=system:admin
+++ oc whoami -t
error: no token is currently in use for this session
++ export IN_CLUSTER_REGISTRY_SECRET_KEY=
++ IN_CLUSTER_REGISTRY_SECRET_KEY=
```
@ksimon1 @0xFelix Do you know how to resolve this issue? because locally I'm unable to reproduce it, as it works without an error.

**Comment by codingben** [comment]
> @0xFelix @ksimon1 This [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/900b3a103dfc4abb373a712919450c6ee1edd83c..c6906fca2c968be16be011b295adbda947b9f464) will check if the containerDisk exists in the container registry. Can you please review?

Seems like this works only on Quay registry and won't work in OpenShfit because API is not the same [1].

[1] https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/image_apis/image-apis

**Comment by codingben** [comment]
> > @0xFelix @ksimon1 This [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/900b3a103dfc4abb373a712919450c6ee1edd83c..c6906fca2c968be16be011b295adbda947b9f464) will check if the containerDisk exists in the container registry. Can you please review?
> 
> Seems like this works only on Quay registry and won't work in OpenShfit because API is not the same [1].
> 
> [1] https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/image_apis/image-apis

This [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/c6906fca2c968be16be011b295adbda947b9f464..dfabcef2e5edbff42f6a7be51aa4f8d3e6a9a980) will resolve this issue and will check if the containerDisk was pushed successfully.

File: modules/disk-uploader/pkg/image/image.go [hunk 14]
```diff
@@ -60,15 +61,20 @@ func Build(diskPath string, config v1.Config) (v1.Image, error) {
 	return image, nil
 }
 
-func Push(image v1.Image, imageDestination string, pushTimeout int) error {
-	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*time.Duration(pushTimeout))
+func Push(image v1.Image, imageDestination string, pushTimeout string) error {
+	timeout, err := strconv.Atoi(pushTimeout)
```
Comment (ksimon1) [inline modules/disk-uploader/pkg/image/image.go:65]: nit: I would add this to Init function in modules/disk-uploader/pkg/parse/clioptions.go file

File: modules/disk-uploader/pkg/parse/clioptions_test.go [hunk 15]
```diff
@@ -39,7 +39,7 @@ var _ = Describe("CLIOptions", func() {
 				ExportSourceName:      expectedExportSourceName,
 				VolumeName:            expectedVolumeName,
 				ImageDestination:      expectedImageDestination,
-				PushTimeout:           60,
+				PushTimeout:           "60",
```
Comment (ksimon1) [inline modules/disk-uploader/pkg/parse/clioptions_test.go:42]: please set timeout to max 10 minutes, so we don't wait for hour in tests

File: test/disk_uploader_test.go [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+		Expect(registryKeyId).ToNot(BeEmpty())
+
+		registryKey := os.Getenv("IN_CLUSTER_REGISTRY_SECRET_KEY")
+		Expect(registryKey).ToNot(BeEmpty())
+
+		registryCredentials, err := createRegistryCredentials(f.CoreV1Client, f.DeployNamespace, registryKeyId, registryKey)
+		Expect(err).ToNot(HaveOccurred())
+
+		cirrosDataVolume := newCirrosDataVolume()
+		cirrosVm, err := createCirrosVM(f.KubevirtClient, f.DeployNamespace, cirrosDataVolume)
```
Comment (ksimon1) [inline test/disk_uploader_test.go:54]: please add `f.ManageVMs(cirrosVm)`, so the vm is deleted at the end of the test

File: test/disk_uploader_test.go [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+		registryKeyId := os.Getenv("IN_CLUSTER_REGISTRY_ACCESS_KEY_ID")
+		Expect(registryKeyId).ToNot(BeEmpty())
+
+		registryKey := os.Getenv("IN_CLUSTER_REGISTRY_SECRET_KEY")
+		Expect(registryKey).ToNot(BeEmpty())
+
+		registryCredentials, err := createRegistryCredentials(f.CoreV1Client, f.DeployNamespace, registryKeyId, registryKey)
+		Expect(err).ToNot(HaveOccurred())
+
+		cirrosDataVolume := newCirrosDataVolume()
```
Comment (ksimon1) [inline test/disk_uploader_test.go:53]: please add `f.ManageDataVolumes(cirrosDataVolume)`, so the DV is deleted after the test

File: modules/disk-uploader/pkg/parse/clioptions.go [hunk 18]
```diff
@@ -18,7 +18,7 @@ type CLIOptions struct {
 	ExportSourceName      string `arg:"--export-source-name" help:"Name of the export source"`
 	VolumeName            string `arg:"--volumename" help:"Name of the volume (if source kind is 'pvc', then volume name is equal to source name)"`
 	ImageDestination      string `arg:"--imagedestination" help:"Destination of the image in container registry"`
-	PushTimeout           int    `arg:"--pushtimeout" help:"Push timeout of container disk to registry"`
+	PushTimeout           string `arg:"--pushtimeout" help:"Push timeout of container disk to registry"`
```
Comment (akrejcir) [inline modules/disk-uploader/pkg/parse/clioptions.go:21]: Why did you change the timeout to string?

Comment (codingben) [reply modules/disk-uploader/pkg/parse/clioptions.go:21]: Tekton expects only string parameters and there is no integer types.

File: modules/disk-uploader/pkg/image/image.go (hunk 14 above)
Comment (codingben) [reply modules/disk-uploader/pkg/image/image.go]: +1

File: test/disk_uploader_test.go (hunk 16 above)
Comment (codingben) [reply test/disk_uploader_test.go:54]: Yep sure!

File: test/disk_uploader_test.go (hunk 17 above)
Comment (codingben) [reply test/disk_uploader_test.go:53]: Absolutely.

**Comment by codingben** [comment]
@ksimon1 I've resolved your comments in the previous revisions. Can you please take a look? functional tests passed locally.

**Comment by codingben** [comment]
@ksimon1 Rebased and there is a new [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/5df1dedc12a65b33027e99b87cf48677692f329a..eb6a00121f7a61e17aeefb5336d5ddb39036fb58) to get back using a normal container registry (could be Quay).

This PR is ready to be merged and functional tests passed locally. I'm putting this PR on draft to disable CI until an external container registry credentials are set, @ksimon1 please advice where they can be configured?

**Comment by codingben** [comment]
This PR depends on https://github.com/openshift/release/pull/59236 currently.

**Comment by codingben** [comment]
Hi @ksimon1 @dhiller, I don't have an access to the Quay repository and credentials you've created and added to the OpenShift CI.

Can you please review them? because it failed on this error:
```
  2024/11/28 10:50:21 Error pushing image: GET https://quay.io/v2/auth?scope=repository%3Akubevirtci%2Ftekton-tasks%3Apush%2Cpull&service=quay.io: unexpected status code 429 Too Many Requests: Too many login attempts. 
  Please reset your Quay password and try again.
```

File: test/disk_uploader_test.go [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+		runner.NewTaskRunRunner(f, config.GetTaskRun()).
+			CreateTaskRun().
+			ExpectSuccess().
+			ExpectLogs(config.GetAllExpectedLogs()...).
+			ExpectResults(nil)
+
+		ref, err := name.ParseReference(imageDestination)
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = remote.Get(ref, remote.WithAuth(&authn.Basic{
```
Comment (0xFelix) [inline test/disk_uploader_test.go:92]: Does the task output an image id or hash? Maybe we can assert that way that the image was updated on the registry?

Comment (ksimon1) [reply test/disk_uploader_test.go:92]: The task does not write any result (because `crane.Push` returns only error and nothing else).  Hower we might take and return the image digest from the variable `containerImage` in this step: https://github.com/kubevirt/kubevirt-tekton-tasks/blob/main/cmd/disk-uploader/main.go#L98 (I am not sure if it is populated during build - needs to be tested if it is not empty) and then check it
```
descriptor, err = remote.Get(ref, remote.WithAuth(&authn.Basic{
...
image := descriptor.Image()
Expect(image.Digest).To(Equal(result.Digest))
```

Comment (codingben) [reply test/disk_uploader_test.go:92]: > Does the task output an image id or hash? Maybe we can assert that way that the image was updated on the registry?

@0xFelix I'm not sure again if the assertion is needed. When you push an image using `go-containerreigstry` it checks if the image was pushed by checking the response status [1]. It will return an error if it failed to push an image to the container registry.

[1] https://github.com/google/go-containerregistry/blob/main/pkg/v1/remote/write.go#L590

> The task does not write any result (because crane.Push returns only error and nothing else).

@ksimon1 I don't understand how the go code inside disk-uploader could update the Tekton Task's result with the image digest, do you have an example for it? besides collecting logs.

Anyways, in your code's example can be added as an extra step after completing the upload of the image to the container registry. This way we can also fail the task's execution from inside execution here [1].

[1] https://github.com/kubevirt/kubevirt-tekton-tasks/blob/main/cmd/disk-uploader/main.go#L109

Comment (ksimon1) [reply test/disk_uploader_test.go:92]: Please check how other tasks are returning results: https://github.com/kubevirt/kubevirt-tekton-tasks/blob/main/cmd/create-vm/main.go#L59

Comment (codingben) [reply test/disk_uploader_test.go:92]: @ksimon1 I'll take a look.

Comment (codingben) [reply test/disk_uploader_test.go]: Done.

**Comment by codingben** [comment]
@0xFelix @ksimon1 This [revision](https://github.com/kubevirt/kubevirt-tekton-tasks/compare/fb8c7b740cc816559248e4c91d990cd67435bb22..0477d18796883653209be8a74422a4e1cffecfc1) will check the image digest given from the build and what was pushed to the container registry.

File: cmd/disk-uploader/main.go [hunk 20]
```diff
@@ -122,26 +122,36 @@ func main() {
...
+		exit.ExitOrDieFromError(constants.GenericExitCode, err)
 	}
 
 	virtClient, err := kubecli.GetKubevirtClient()
 	if err != nil {
-		exit.ExitOrDieFromError(genericExitCode, err)
+		exit.ExitOrDieFromError(constants.GenericExitCode, err)
+	}
+
+	image, err := run(*cliOptions, k8sClient, virtClient)
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply cmd/disk-uploader/main.go:144]: Let's do it in a follow up.

File: modules/disk-uploader/pkg/constants/constants.go [hunk 21]
```diff
@@ -0,0 +1,12 @@
+package constants
```
[the remark this answers is not in the document]
Comment (ksimon1) [reply modules/disk-uploader/pkg/constants/constants.go:1]: This constants file is following pattern, which other modules have. We can remove it in follow up PR. I will create an issue for it

[the remark this answers is not in the document]
Comment (0xFelix) [reply modules/disk-uploader/pkg/constants/constants.go:1]: Ok for now as it follow how it is done in the other modules.
